### PR TITLE
Add LBFGS optimization algorithm to C++ API

### PR DESCRIPTION
**Pull request by lematt1991**
This PR ports the Python implementation of [LBFGS](https://github.com/pytorch/pytorch/blob/master/torch/optim/lbfgs.py) to C++.  Additionally, it adds the [closure idiom](https://pytorch.org/docs/stable/optim.html#optimizer-step-closure) that exists in the Python API (required for LBFGS).

### CODE REVIEW

**Comment by goldsborough** [comment]
@ebetica  can you take a look as this touches mostly autogradpp code I haven't worked with yet?

File: torch/csrc/api/include/torch/optimizers.h [hunk 1]
```diff
@@ -13,6 +13,7 @@ class OptimizerImpl {
   virtual ~OptimizerImpl() = default;
   virtual void init_state() {}
   virtual void step() = 0;
+  virtual at::Scalar step(std::function<at::Scalar()> closure) = 0;
```
Comment (ebetica) [inline torch/csrc/api/include/torch/optimizers.h]: I think I prefer just one function, but where the signature looks like:

```c++
virtual at::Scalar step(std::function<at::Scalar()> closure = OptimizerImpl::NoLoss) = 0;
```

and define NoLoss to just return an empty scalar. Keep in mind you'll have to define the default again in each of the subclasses as well. It's slightly cleaner than having a separate step() function that users might have to remember to implement.

File: torch/csrc/api/src/optimizers.cpp [hunk 2]
```diff
@@ -18,6 +18,165 @@ void OptimizerImpl::set_model(std::shared_ptr<nn::Module> model) {
...
+  if(at::Scalar(abs_grad_sum).toFloat() <= tolerance_grad_) {
+    return loss;
+  }
+
+  at::Tensor ONE = flat_grad.type().scalarTensor(1);
+
+  int n_iter = 0;
+  while(n_iter < max_iter_) {
+    n_iter++;
+    state_n_iter += 1;
```
Comment (ebetica) [inline torch/csrc/api/src/optimizers.cpp]: nit: state_n_iter++

File: torch/csrc/api/src/optimizers.cpp [hunk 3]
```diff
@@ -18,6 +18,165 @@ void OptimizerImpl::set_model(std::shared_ptr<nn::Module> model) {
...
+
+    if(state_n_iter == 1) {
+      d = flat_grad.neg();
+      H_diag = flat_grad.type().scalarTensor(1);
```
Comment (ebetica) [inline torch/csrc/api/src/optimizers.cpp]: nit: H_diag = ONE;

**Review by ebetica — COMMENTED** [review]
Looks good as long as the tests pass and nits are fixed. @goldsborough you might want to see if the way we do step is okay here. I think it's basically how pytorch would do it.

**Comment by lematt1991** [comment]
It seems that the LBFGS test is a bit flaky.  I recreated the same thing in Python and am getting some flaky-ness from the [Python version](https://gist.github.com/lematt1991/6c74ff9bfd0063cbaa6d94ac4ed0f064) as well.  Any suggestions on how to make this more deterministic?

**Comment by ezyang** [comment]
@lematt1991 The very least you can do is make sure all of the RNGs are being seeded deterministically at the beginning of your test. Another possibility is to compare the C++ algo against the Python one instead of testing for convergence; you're less likely to have misimplemented both of them. Are there some mathematical properties that LBFGS optimization should have? Test for those.

**Review by ebetica — APPROVED** [review]
Thanks @lematt1991 !

**Comment by goldsborough** [comment]
I think your test is flaky https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-trusty-py3.6-gcc7.2-test/6358/console

**Comment by lematt1991** [comment]
Hmm, that's strange.  I've pulled the `linux-trusty-py3.6-gcc7.2` docker container and tried to reproduce this issue.  If I print out the `running_loss` in each iteration and dump it to a file, multiple runs are giving me the exact same results:

```
jenkins@c8c8df3924a6:/workspace$ ./tools/cpp_build/build/libtorch/bin/test_api optim --section lbfgs > out1
CUDA not available. Disabling CUDA tests
jenkins@c8c8df3924a6:/workspace$ ./tools/cpp_build/build/libtorch/bin/test_api optim --section lbfgs > out2
CUDA not available. Disabling CUDA tests
jenkins@c8c8df3924a6:/workspace$ head -n 5 out1
loss = 0.996719
loss = 0.994648
loss = 0.991368
loss = 0.988453
loss = 0.985467
jenkins@c8c8df3924a6:/workspace$ diff out1 out2
jenkins@c8c8df3924a6:/workspace$
```

I've seen that LBFGS is very sensitive as even very minor changes to any of the hyper parameters seem to cause it to diverge.  Maybe following @ezyang's suggestion of testing it by comparing it to the Python implementation would be a better approach, but I'm not sure I see what could be causing the nondeterminism with all of the RNGs seeded.  What are your guys' thoughts?

**Pull request merged**